rdmd_test: rewrite to use externally-specifed rdmd executable

WebDrake opened the pull request:
One key problem of the old `rdmd_test` was that it built its own rdmd instance to test.  This meant that technically, it was never actually run against the rdmd executable generated in the regular tools build.

This rewrite fixes this by repurposing the `--rdmd` flag of `rdmd_test` to require not a path to `rdmd.d` but a path to the rdmd executable for which the test suite should be run.  The provided executable is not used directly, but is copied into a temporary directory: this is necessary in order to run the fallback test, which involves copying a dummy compiler into the same directory as the tested rdmd executable.

The `--compiler` flag has been removed (since `rdmd_test` no longer uses it to build an rdmd executable), but a new `--fallback-compiler` flag is now required in order to specify the name of the compiler to use in the fallback test.  Note that this test will fail if the provided name does not match the name of the default compiler used by the provided rdmd.

The `test_rdmd` target in `posix.mak` has been reworked to use the new `--rdmd` and `--fallback-compiler` flags, whose values can be specified with the `RDMD_TEST_EXECUTABLE` and `RDMD_TEST_FALLBACK_COMPILER` make variables respectively.  By default, these will be set to the generated `rdmd` instance and the compiler used to build it.

This should ensure that the rdmd instance generated by the regular build is properly tested, while allowing `rdmd_test` to be used with arbitrary `rdmd` instances if any developer should wish to do so.

The first and last patches perform a bit of cleanup useful for the main changes.

Note that an assumption made by this PR is that breaking changes to `rdmd_test` are not a concern, since this is a test suite rather than a utility intended for distribution.

CyberShadow commented:
This seems unnecessarily complicated. Why the complexity, and why break existing workflows? Why not just skip compiling rdmd in the presence of a new flag? Or if `--rdmd` points to a non-`.d`-file?

> Note that an assumption made by this PR is that breaking changes to rdmd_test are not a concern, since this is a test suite rather than a utility intended for distribution.

I think it would be nice to keep the behaviour of building `rdmd` by default, as it avoids accidentally testing an `rdmd` binary that doesn't correspond to the source code. Mismatch in compiler flags seems less likely to lead to frustration.

WebDrake commented:
Very good questions.  In fact the original implementation I put together followed more that path -- of adding a new `--rdmd-executable` flag, and not removing or repurposing any existing ones.

These are the reasons I abandoned that approach:

* `rdmd_test` is only built via the `test_rdmd` and `test` targets in `posix.mak`: it's a test utility only, not distributed as part of the tools

* in consequence, the only breaking use-case is that of someone hacking on rdmd who prefers to directly invoke the `rdmd_test` executable instead of `make -f posix.mak test_rdmd`

  * (... and given that the total number of devs is small, this seems likely to be an even smaller number ...)

* for anyone who uses `make -f posix.mak test{_rdmd}`, the _effective_ behaviour has not changed at all: any change to `rdmd.d` will result in a rebuild that will then be tested

* the only way to accidentally test the 'wrong' `rdmd` is to explicitly provide it via the makefile variable -- if you don't do this, the generated `rdmd` will automatically be used

* doing things this way creates a clean separation between _building_ rdmd (which is now 100% the responsibility of the makefile) and _testing_ rdmd

* this means that it is now impossible (via the standard build process) to avoid testing the actual rdmd that will be used, as opposed to some built-by-other-means rdmd that never gets used outside of the test suite

Note that it's possible to strip out the separation of concerns still further: one could strip out the `--test-compilers`, just have `--rdmd` and `--compiler` flags, and have a boolean `--test-fallback` option, so that the choice to iterate over all the available D compilers would be made entirely on the CI side of things.

Anyway, I'm happy to consider a rewrite, but it would be good to have thoughts and feedback on the above set of points before deciding how to proceed.

CyberShadow commented:
> the only way to accidentally test the 'wrong' rdmd is to explicitly provide it via the makefile variable -- if you don't do this, the generated rdmd will automatically be used

Sorry, how so? What happens if someone:

1. Clones the tools repo
2. Runs `make`
3. Changes something in `rdmd.d`
4. Runs `dmd -run rdmd_test.d`

Perhaps an alternative way to fix this would be, instead of `rdmd_test` building `rdmd` using a `dmd` invocation, to instead invoke `make`?

WebDrake commented:
> Runs dmd -run rdmd_test.d

But why would they do that, instead of just running `make test_rdmd`?

WebDrake commented:
... or even just `make test`.  The key thing here is that the very _existence_ of `rdmd_test` really ought to be an under-the-hood implementation detail to anyone hacking on rdmd.

CyberShadow commented:
> But why would they do that, instead of just running make test_rdmd?

Well, I can mostly speak of this from my personal viewpoint, but in my text editor, I hit a key and it executes and runs the current file, and in my file manager, I hit another key and it builds/runs the file under the cursor.

Building/running a D program the usual way D programs are built/run is the most discoverable way to do this, because it doesn't require discoverability - it relies on common prior knowledge. So, I think it would not be unexpected that others would do the same.

WebDrake commented:
Well, I certainly see why you find that useful.  At the same time, it sounds like that approach might be achieving one version of simplicity at the expense of a genuinely robust approach.

There's a reasonable point of view that the only valid way to test rdmd (and indeed any of the programs in the tools repo) is to run `make test`.  From this point of view building and running `rdmd_test` directly is a bit like trying to heat your apartment by lighting the gas hob on your cooker -- yes, it will generate immediate heat, but it's not actually doing what is needed in an appropriately effective way.

Bearing in mind that (i) it's really easy to understand that a project uses `make` and (ii) it's really easy to set up an IDE to invoke `make` for given targets, I'm not sure that it's a good idea to encourage or support developers in taking non-standard approaches to building and development.

Obviously it's nice if we don't block things that you (or other developers) find useful, but I wouldn't privilege that way of doing things over ensuring a good separation of concerns between build and test responsibilities.

CyberShadow commented:
> non-standard approaches to building and development

Err, no, there's nothing non-standard about running a D program the standard way you run a D program.

A lot of well-designed test suites allow running individual test scripts directly in addition to the higher-level build system targets. Fine, I'll accept that there can be good arguments to not support this approach here for `rdmd_test` (though I don't think they have been presented so far), but, if so, then attempting to use this "wrong" approach must fail loudly and clearly and point the user to the "correct" approach (i.e. if some switch to `rdmd_test` is missing, tell the user to use the `make` target instead of attempting to invoke the test script directly.)

But, really, it should just work and just do the right thing. Why not invoke `make` as I proposed above?

WebDrake commented:
> Err, no, there's nothing non-standard about running a D program the standard way you run a D program.

Obviously one can always invoke a compiler directly (and not just for D) and run the resulting executable.  But one then has to take personal responsibility for invoking both compiler and executable correctly.

> A lot of well-designed test suites allow running individual test scripts directly in addition to the higher-level build system targets.

Yes, and that's still allowed here.  What has changed is that there is now a robust separation of concerns between building the thing that is being tested, versus running the test script.

> But, really, it should just work and just do the right thing.

Well, technically it will, if one just does `make test` ... or, if one wants to invoke the individual test script, `make test_rdmd`.

> Why not invoke `make` as I proposed above?

For starters, because `rdmd_test` cannot reasonably know where `make` will place the generated rdmd executable.  We could second-guess it, but then it involves duplication between `rdmd_test` and the makefile ... which is why it is much easier to just say that the makefile is responsible for coordinating builds and tests, whereas `rdmd_test` is responsible for testing any given `rdmd` executable instance you care to provide it with.

CyberShadow commented:
> But one then has to take personal responsibility for invoking both compiler and executable correctly.

"Personal responsibility" sounds like a convenient way to say that we just don't want to help users in avoiding them shooting themselves in the foot. It is not a constructive way to approach things.

If we want to go this way, then it should be made blatantly clear that running `rdmd_test` directly is not the intended way to use it.

- The documentation at the top of the file should be updated.
- `rdmd_test` should mention the correct way to use it when ran without arguments.
- If we're shifting towards the makefile as the one and only way to do things, then some documentation in the makefile, similar to other projects' makefiles ([dlang.org](https://github.com/dlang/dlang.org/blob/master/posix.mak#L1), [dmd](https://github.com/dlang/dmd/blob/master/src/posix.mak#L1)), would be nice.

Most importantly, this is missing corresponding changes in win32.mak, as the old method to run rdmd tests on Windows won't work any more.

WebDrake commented:
> "Personal responsibility" sounds like a convenient way to say that we just don't want to help users in avoiding them shooting themselves in the foot. It is not a constructive way to approach things.

You're speaking about this as if this is some widely-used, widely-distributed application.  But it's not: it's a test suite with a very narrow user base, _that is never built outside of the `make test_{rdmd}` targets_, and it's shooting _everyone_ in the foot if it tests anything other than `rdmd` as it is built by the standard build scripts.

On the other hand,

> If we want to go this way, then it should be made blatantly clear that running `rdmd_test` directly is not the intended way to use it.

... it's perfectly possible to run `rdmd_test` directly if one really wants to, but in this case one has to supply the `rdmd` instance to test, and information about the fallback compiler it is expecting.

What is _not_ reasonable is to expect that just invoking `rdmd_test` will actually magically validate the `rdmd` source code, because `rdmd_test` has no reasonable way to know what is the proper way to actually build it.  (YMMV, but right now if I just try to run the master-branch `./rdmd_test.d`, taking advantage of the shebang, it fails with an error anyway, admittedly not when building the tested rdmd instance.)

Anyway, to your specific suggestions:

> The documentation at the top of the file should be updated.

Yup, quite right.  Sorry for missing it.

> `rdmd_test` should mention the correct way to use it when ran without arguments.

I'll see what I can do here.  Probably the best way is to use `getopt`'s support for help documentation.

> If we're shifting towards the makefile as the one and only way to do things, then some documentation in the makefile, similar to other projects' makefiles (dlang.org, dmd), would be nice.

Agree, but that should probably be a PR in its own right (which I'm happy to make), after the current work is finalized.

CyberShadow commented:
Agreed with all of the above, thanks! Please ping when you push the doc updates.

WebDrake commented:
Will do.  One quick question:

> Most importantly, this is missing corresponding changes in win32.mak, as the old method to run rdmd tests on Windows won't work any more.

Can you clarify what you mean here?  AFAIC `rdmd_test` is neither built nor invoked by any `win32.mak` target.  Is there some other mechanism for testing Windows builds of rdmd?

I assume there's nothing, because otherwise tests would fail, and they don't ... :-\

CyberShadow commented:
> Can you clarify what you mean here? AFAIC rdmd_test is neither built nor invoked by any win32.mak target. Is there some other mechanism for testing Windows builds of rdmd?

You used to just run `rdmd_test`...

I don't think it's hooked into any Windows CI right now. Digger knows how to test rdmd, but it's not hooked up as a CI either.

WebDrake commented:
> You used to just run `rdmd_test`...

Ah, then I really apologize; I've misunderstood the whole basis for your remarks about workflows.  Killing people's ability to work on Windows is unacceptable (although arguably, so too is the fact that CI does not run `rdmd_test` for Windows builds...).

I'll see if I can update `win32.mak` to ensure things work there too.

WebDrake commented:
I've updated the PR with a number of changes and improvements:

* the `posix.mak` cleanup commit now comes before all other changes (simplifies later diffs)

* some small tweaks have been made to the main rework-`rdmd_test` patch, notably the `--fallback-compiler` flag has been replaced by `--rdmd-default-compiler` and the corresponding make variable is `RDMD_TEST_DEFAULT_COMPILER`.

* `rdmd_test.d` module doc has been updated to reflect the new behaviour and usage

* I've added `test_rdmd` and `test` targets to `win32.mak` and updated `appveyor.yml` to use them

Still TBD: need to update how `rdmd_test` behaves if either given no arguments or if given nothing.

WebDrake commented on appveyor.yml:
```diff
@@ -64,3 +64,4 @@ test_script:
  - echo %PATH%
  - '%DC% --version'
  - make -f win32.mak rdmd dustmite ddemangle changed
+ - make -f win32.mak test
```
I should add `VERBOSE_RDMD_TEST=1` here so that we actually see the tests being performed.

wilzbach commented:
> I've added test_rdmd and test targets to win32.mak and updated appveyor.yml to use them

:+1: - thanks!

WebDrake commented:
N.B. wait for me to squash the `fixup!` patch, assuming it works ;-)

WebDrake commented:
Is it usual for appveyor to take this long even to start a build? :-\

wilzbach commented:
Yeah, we only have one parallel job per account in the free version and DMD has quite a lot of traffic. The quick workaround would be to use a different account like we have done for `dlang/installer`.
I personally just use Wine to test out things, since 2.078 it works quite nicely.

WebDrake commented:
OK, now I'm mystified (especially by this `line 117` thing, since there aren't even 117 lines in the file).  Does anyone more experienced with the windows makefile have any suggestions?

wilzbach commented on win32.mak:
```diff
@@ -86,3 +86,28 @@ zip: detab tolf $(MAKEFILES)
+VERBOSE_RDMD_TEST=0
+ifeq ($(VERBOSE_RDMD_TEST), 1)
+	override VERBOSE_RDMD_TEST_FLAGS:=-v
+endif
```
I think this isn't supported by DigitalMars make

WebDrake replied:
Ho hum.  Probably best to just specify `VERBOSE_RDMD_TEST_FLAGS` directly then.

WebDrake replied:
Or just always pass `-v`, which is probably desirable.

WebDrake commented:
Well, this works, but is it intended that we only have a 32-bit build?  There seems to be no `MODEL` parameter in the makefile and `generated` files are sent to the `32` directory.

CyberShadow commented:
Currently we don't ship 64-bit Windows executables, so I guess the answer would be "yes". Although a win64.mak akin to that in the other D repos could be added, hopefully there's not much Windows-specific AND 64-bit-specific in rdmd that would benefit from that test coverage.

WebDrake commented:
> Currently we don't ship 64-bit Windows executables, so I guess the answer would be "yes".

I should clarify: the concern isn't about the build of `rdmd.exe` but about the `-m` flag that the tested `rdmd` provides to the compiler.

So, if on x64 we should expect a 64-bit `dmd` executable, which by default would produce 64-bit code, we should probably validate that use-case.

If I understand the Digital Mars make docs correctly (https://www.digitalmars.com/ctg/make.html) then I think I can use `if` conditions to ensure the correct behaviour here.  I'll push a fixup patch to try this out.

WebDrake commented:
OK, so that doesn't work.  For now then I'll just hardcode `-m32` and leave this as a follow-up task.

WebDrake commented:
OK, I've squashed everything down and added a couple of key changes:

* the refactoring patch now includes help usage info via the `getopt` call

* I've added a final patch to add the `MODEL=32,64` parameter via the appveyor.yml matrix.  If this doesn't work it can be dropped, but hopefully it will allow proper testing on the target platform.

WebDrake commented:
Tests are looking good, I think this is ready for final review, folks :-)

CyberShadow commented:
Great work! This looks pretty much perfect.

Did you decide to not add the mention of the makefile target to the rdmd_test getopt help text after all?

WebDrake commented:
> Did you decide to not add the mention of the makefile target to the rdmd_test getopt help text after all?

Yes.  I thought that if the user gets to the point of actually running it, they probably just want to be informed on how it works in its own right.  But I can modify if necessary (e.g. if we find that people are getting confused about how it's supposed to work).

wilzbach commented on appveyor.yml:
```diff
@@ -4,9 +4,11 @@ environment:
   - DC: dmd
     DVersion: nightly
     arch: x64
+    MODEL: 64
```
FWIW this works, because it has no effect. Though fixing this goes over the scope of this PR.

WebDrake replied:
What do you mean "it has no effect"?  It's explicitly used when invoking `rdmd_test.exe` and determines how `rdmd` invokes the underlying D compiler.

I'm aware it doesn't affect any part of how the tools are built.

wilzbach replied:
Sorry didn't see that line from my phone.

WebDrake replied:
OK, no worries.  Just wanted to make sure there was nothing unexpected in how `dmd` on Win64 would react to the `-m` flag.

wilzbach replied:
Looks all good:

```
rdmd=generated\windows\32\rdmd.exe -m64 -v  --rdmd-default-compiler=dmd  --test-compilers=dmd
[execute] ["C:\\Users\\appveyor\\AppData\\Local\\Temp\\1\\rdmd_app_.exe"]
[execute] ["C:\\Users\\appveyor\\AppData\\Local\\Temp\\1\\rdmd_app_.exe", "--help"]3[execute] ["C:\\Users\\appveyor\\AppData\\Local\\Temp\\1\\rdmd_app_.exe", "--compiler=dmd", "-m64", "C:\\Users\\appveyor\\AppData\\Local\\Temp\\1\\force_src_.d"]
74[execute] 
...
```

wilzbach commented on rdmd_test.d:
```diff
@@ -46,40 +50,55 @@ bool verbose = false;
-        "v|verbose", &verbose,
+    auto helpInfo = getopt(args,
+        "rdmd", "[REQUIRED] path to rdmd executable to test", &rdmd,
+        "rdmd-default-compiler", "[REQUIRED] default D compiler used by rdmd executable", &defaultCompiler,
```
getopt supports `config.required`

CyberShadow replied:
Fairly sure a previous version of this PR used it. I guess it was taken out to allow displaying a better error message?

wilzbach commented on rdmd_test.d:
```diff
@@ -46,40 +50,55 @@ bool verbose = false;
 
-    auto res = execute([buildCompiler, modelSwitch(model), "-of" ~ rdmdApp, rdmd]);
+    if (defaultCompiler.length == 0)
+        reportHelp("ERROR: missing required --rdmd-default-compiler flag");
```
getopt supports this out of the box.

WebDrake replied:
I'm aware of `std.getopt.config.required`, but I found the behaviour really undesirable (it was very unhelpful in its error messages).  So it seemed better to do something small, manual and actually helpful in its output.

wilzbach replied:
Hmm, I guess we have to improve std.getopt then at some point...
